### PR TITLE
instance norm and reflection padding

### PR DESCRIPTION
**Pull request by zhanghang1989**
Re-create the PR for instance norm and reflection padding.

### CODE REVIEW

**Comment by szha** [comment]
Could you fix lint? The errors can be obtained by running `make pylint`

File: python/mxnet/gluon/nn/basic_layers.py [hunk 1]
```diff
@@ -442,3 +442,94 @@ def hybrid_forward(self, F, x):
...
+        initialization will be deferred to the first time `forward` is called
+        and `in_channels` will be inferred from the shape of input data.
+
+    Input shape:
+        Arbitrary.
+
+    Output shape:
+        Same shape as input.
+
+       This implementation is based on paper:
```
Comment (piiswrong) [inline python/mxnet/gluon/nn/basic_layers.py]: fix format. see dropout

File: python/mxnet/gluon/nn/conv_layers.py [hunk 2]
```diff
@@ -1009,3 +1009,32 @@ def __init__(self, layout='NCDHW', **kwargs):
...
         super(GlobalAvgPool3D, self).__init__(
             (1, 1, 1), None, 0, True, True, 'avg', **kwargs)
+
+
+class ReflectionPad2D(HybridBlock):
+    """Pads the input tensor using the reflection of the input boundary.
+
+    Parameters
+    ----------
+        pad_width int: the size of the padding. If is int, uses the same
```
Comment (piiswrong) [inline python/mxnet/gluon/nn/conv_layers.py]: fix format

File: python/mxnet/gluon/nn/conv_layers.py [hunk 3]
```diff
@@ -1009,3 +1009,32 @@ def __init__(self, layout='NCDHW', **kwargs):
...
+          :math:`H_{out} = H_{in} + paddingTop + paddingBottom`
+          :math:`W_{out} = W_{in} + paddingLeft + paddingRight`
+
+    Examples
+    --------
+    >>> m = nn.ReflectionPad(3)
+    >>> input = mx.nd.random_normal(shape=(16, 3, 224, 224))
+    >>> output = m(input)
+    """
+    def __init__(self, pad_width=0, **kwargs):
```
Comment (piiswrong) [inline python/mxnet/gluon/nn/conv_layers.py]: pad_width -> padding

File: python/mxnet/gluon/nn/conv_layers.py (hunk 2 above)
Comment (piiswrong) [reply python/mxnet/gluon/nn/conv_layers.py]: if not int what else can it be?

File: python/mxnet/gluon/nn/conv_layers.py [hunk 4]
```diff
@@ -1009,3 +1009,32 @@ def __init__(self, layout='NCDHW', **kwargs):
...
+
+class ReflectionPad2D(HybridBlock):
+    """Pads the input tensor using the reflection of the input boundary.
+
+    Parameters
+    ----------
+        pad_width int: the size of the padding. If is int, uses the same
+            padding in all boundaries. 
+
+    Shape:
```
Comment (piiswrong) [inline python/mxnet/gluon/nn/conv_layers.py]: fix format

File: python/mxnet/gluon/nn/conv_layers.py [hunk 5]
```diff
@@ -1009,3 +1009,32 @@ def __init__(self, layout='NCDHW', **kwargs):
...
+        - Input: :math:`(N, C, H_{in}, W_{in})`
+        - Output: :math:`(N, C, H_{out}, W_{out})` where
+          :math:`H_{out} = H_{in} + paddingTop + paddingBottom`
+          :math:`W_{out} = W_{in} + paddingLeft + paddingRight`
+
+    Examples
+    --------
+    >>> m = nn.ReflectionPad(3)
+    >>> input = mx.nd.random_normal(shape=(16, 3, 224, 224))
```
Comment (piiswrong) [inline python/mxnet/gluon/nn/conv_layers.py]: mx.nd.random.normal

File: python/mxnet/gluon/nn/conv_layers.py [hunk 6]
```diff
@@ -1009,3 +1009,32 @@ def __init__(self, layout='NCDHW', **kwargs):
...
+
+    Shape:
+        - Input: :math:`(N, C, H_{in}, W_{in})`
+        - Output: :math:`(N, C, H_{out}, W_{out})` where
+          :math:`H_{out} = H_{in} + paddingTop + paddingBottom`
+          :math:`W_{out} = W_{in} + paddingLeft + paddingRight`
+
+    Examples
+    --------
+    >>> m = nn.ReflectionPad(3)
```
Comment (piiswrong) [inline python/mxnet/gluon/nn/conv_layers.py]: m -> layer

File: python/mxnet/gluon/nn/conv_layers.py [hunk 7]
```diff
@@ -1009,3 +1009,32 @@ def __init__(self, layout='NCDHW', **kwargs):
...
+    >>> input = mx.nd.random_normal(shape=(16, 3, 224, 224))
+    >>> output = m(input)
+    """
+    def __init__(self, pad_width=0, **kwargs):
+        super(ReflectionPad2D(, self).__init__(**kwargs)
+        self._pad_width = pad_width
+        
+    def forward(self, x):
+        return F.pad(x, mode='reflect', pad_width=self._pad_width)
```
Comment (piiswrong) [inline python/mxnet/gluon/nn/conv_layers.py]: shouldn't pad_width be a tuple?

**Comment by piiswrong** [comment]
Please add tests.

**Comment by piiswrong** [comment]
@szha The original author seems to have disappeared. Would you like to take over?

**Comment by szha** [comment]
Actually @zhanghang1989 will join us soon 😄

**Comment by zhanghang1989** [comment]
@piiswrong @szha Hi guys, sorry for the delay. I just made some changes.

**Comment by piiswrong** [comment]
see the other ops for doc format.

Also please add test cases

**Comment by zhanghang1989** [comment]
Thanks @piiswrong! I made some changes to the doc.

**Comment by zhanghang1989** [comment]
I will add one unit test soon at unittest/test_gluon.py . Thanks!

**Comment by szha** [comment]
please fix lint

File: python/mxnet/gluon/nn/conv_layers.py [hunk 8]
```diff
@@ -1007,3 +1008,32 @@ def __init__(self, layout='NCDHW', **kwargs):
...
+    >>> m = nn.ReflectionPad2D(3)
+    >>> input = mx.nd.random.normal(shape=(16, 3, 224, 224))
+    >>> output = m(input)
+    """
+    def __init__(self, padding=0, **kwargs):
+        super(ReflectionPad2D, self).__init__(**kwargs)
+        self._padding = padding
+
+    def hybrid_forward(self, F, x):
+        return F.pad(x, mode='reflect', padding=self._padding)
```
Comment (szha) [inline python/mxnet/gluon/nn/conv_layers.py]: This isn't tested yet. Also, it looks like a very straightforward one-liner. Should we just ask users to use hybrid lambda for this?

Comment (zhanghang1989) [reply python/mxnet/gluon/nn/conv_layers.py]: This simple API makes user easier to use. The Reflectance padding is popularly used in low level vision, such as GANs, super resolution and style transfer. I think it make sense to provide an API for it. What's your thought?

http://pytorch.org/docs/0.3.0/nn.html?highlight=padding#reflectionpad2d

Comment (piiswrong) [reply python/mxnet/gluon/nn/conv_layers.py]: I think we should have this.

File: python/mxnet/gluon/nn/conv_layers.py [hunk 9]
```diff
@@ -1007,3 +1008,32 @@ def __init__(self, layout='NCDHW', **kwargs):
...
         super(GlobalAvgPool3D, self).__init__(
             (1, 1, 1), None, 0, True, True, 'avg', **kwargs)
+
+
+class ReflectionPad2D(HybridBlock):
+    """Pads the input tensor using the reflection of the input boundary.
+
+    Parameters
+    ----------
+        `padding` is a tuple of integer padding widths for each axis of the format
```
Comment (piiswrong) [inline python/mxnet/gluon/nn/conv_layers.py]: Please fix the documentation and add tests

File: python/mxnet/gluon/nn/conv_layers.py [hunk 10]
```diff
@@ -1007,3 +1008,32 @@ def __init__(self, layout='NCDHW', **kwargs):
...
+
+    Examples
+    --------
+    >>> m = nn.ReflectionPad2D(3)
+    >>> input = mx.nd.random.normal(shape=(16, 3, 224, 224))
+    >>> output = m(input)
+    """
+    def __init__(self, padding=0, **kwargs):
+        super(ReflectionPad2D, self).__init__(**kwargs)
+        self._padding = padding
```
Comment (piiswrong) [inline python/mxnet/gluon/nn/conv_layers.py]: Are you sure this works correctly when padding is a number instead of a tuple?

File: python/mxnet/gluon/nn/conv_layers.py (hunk 8 above)
Comment (szha) [reply python/mxnet/gluon/nn/conv_layers.py]: OK

File: python/mxnet/gluon/nn/conv_layers.py (hunk 10 above)
Comment (zhanghang1989) [reply python/mxnet/gluon/nn/conv_layers.py]: just updated the integer type check and fix docs

File: python/mxnet/gluon/nn/basic_layers.py [hunk 11]
```diff
@@ -480,6 +480,88 @@ def __repr__(self):
...
+     [[-0.99998319  0.99998361]]]
+    <NDArray 2x1x2 @cpu(0)>
+    """
+    def __init__(self, epsilon=1e-5, center=True, scale=False,
+                 beta_initializer='zeros', gamma_initializer='ones',
+                 in_channels=0, **kwargs):
+        super(InstanceNorm, self).__init__(**kwargs)
+        self._kwargs = {'eps': epsilon}
+        if in_channels != 0:
+            self.in_channels = in_channels
```
Comment (szha) [inline python/mxnet/gluon/nn/basic_layers.py]: This doesn't seem needed in forward. The reference in `__repr__` can be changed to use the shape from weight instead.

File: python/mxnet/gluon/nn/basic_layers.py [hunk 12]
```diff
@@ -480,6 +480,88 @@ def __repr__(self):
...
+                                    allow_deferred_init=True)
+
+    def hybrid_forward(self, F, x, gamma, beta):
+        return F.InstanceNorm(x, gamma, beta,
+                              name='fwd', **self._kwargs)
+
+    def __repr__(self):
+        s = '{name}({content}'
+        if hasattr(self, 'in_channels'):
+            s += ', in_channels={0}'.format(self.in_channels)
```
Comment (szha) [inline python/mxnet/gluon/nn/basic_layers.py]: load size from weight instead.

File: python/mxnet/gluon/nn/conv_layers.py [hunk 13]
```diff
@@ -1007,3 +1008,37 @@ def __init__(self, layout='NCDHW', **kwargs):
...
+
+class ReflectionPad2D(HybridBlock):
+    """Pads the input tensor using the reflection of the input boundary.
+
+    Parameters
+    ----------
+    padding: int or a tuple of int
+        An integer padding width for height and weight or a tuple of
+        integers padding widths for each axis of the format ``(before_1, after_1, ... ,
+        before_N, after_N)``. The `padding` should be of length ``2*N`` where ``N``
```
Comment (szha) [inline python/mxnet/gluon/nn/conv_layers.py]: The document describes something general but the class is specific to 2D. Maybe describe the argument that's specific to 2D case here.

Comment (zhanghang1989) [reply python/mxnet/gluon/nn/conv_layers.py]: If pass into a tuple, it works for different dimensions. Should we remove this?

Comment (szha) [reply python/mxnet/gluon/nn/conv_layers.py]: Since the class name says 2D, it should support 2D.

You can do something similar to the conv blocks, with a common abstract class for the forward logic, and the actual 1D/2D/3D classes for specific implementation.

Comment (zhanghang1989) [reply python/mxnet/gluon/nn/conv_layers.py]: Agree. Just made the changes. Will refactor when adding 1D/3D

File: python/mxnet/gluon/nn/basic_layers.py (hunk 11 above)
Comment (zhanghang1989) [reply python/mxnet/gluon/nn/basic_layers.py]: Actually, BatchNorm uses the same thing
https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/nn/basic_layers.py#L375

File: python/mxnet/gluon/nn/basic_layers.py [hunk 14]
```diff
@@ -555,8 +553,9 @@ def hybrid_forward(self, F, x, gamma, beta):
 
     def __repr__(self):
         s = '{name}({content}'
+        in_channels = self.gamma.shape[0]
         if hasattr(self, 'in_channels'):
```
Comment (szha) [inline python/mxnet/gluon/nn/basic_layers.py]: remove condition

**Comment by piiswrong** [comment]
@zhanghang1989 @szha 
The instance norm operator needs an axis argument to be consistent with batchnorm

reflectionpad2d's input/output shape doc has wrong format
